lib/cylc/mp_pool.py

```diff
@@ -159,7 +163,14 @@ def process(self):
         for proc, ctx, callback, callback_args in self.runnings:
             ret_code = proc.poll()
             if ret_code is None:
-                runnings.append([proc, ctx, callback, callback_args])
+                if time.time() > ctx.init_time + self.cmd_timeout:
+                    os.killpg(proc.pid, SIGKILL)
+                    ctx.ret_code = 1
+                    ctx.err = "killed on timeout (%s)" % self.cmd_timeout
+                    self._run_command_exit(ctx, callback, callback_args)
+                    proc.wait()
```

Comment:
Better this way?

```python
os.killpg(proc.pid, SIGKILL)
ctx.ret_code = proc.wait()
ctx.out, ctx.err = ctx.communicate()
ctx.err += '\nkilled on timeout (%s)' % self.cmd_timeout
self._run_command_exit(ctx, callback, callback_args)
```

`ctx.ret_code` should then be set to the actual exit value. If the program has got some stderr and stdout, then they will be logged as well.